### PR TITLE
Use date library in editor and fix relative date display

### LINKED ISSUE #4095: Dashboard: Created/last modified relative strings are sometimes broken
## Bug Description

The first story was modified a few minutes ago.

## Expected Behaviour

It should print the correct relative time.

## Steps to Reproduce

The screenshot is from staging (I'm in UTC+2).

## Screenshots

image

## Additional Context

- Plugin Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

This is a follow-up to #3247, #3244, and #3186.

We currently use the `@wordpress/date` package for date formatting in the editor.

This causes a couple of issues:

* We load it as an **external** (meaning it is not part of the actual bundle) because it is GPL licensed...
* ...which means it itself loads `moment` as an external since it's a dependency.
* An additional hurdle for decoupling the app from WordPress (#2918)

What we need is a simple way to format arbitrary dates and times in the editor via the existing date utils we have.

## Relevant Technical Choices

* New `formatDate` and `formatTime` utils.

## To-do

* [x] ~Fix `getRelativeDisplayDate` tests~ Fix `isToday`/`isYesterday` so tests pass 
* [x] Add unit tests for new functions
* [ ] Improve API for dates?  
  Comparing with https://github.com/WordPress/gutenberg/tree/master/packages/date I think we still have some room for improvement here.  
  For example, it's cumbersome to pass timezone information every time I want to format a date. There should be a way to set a default using something like `setSettings`.
  -> **probably new issue**
* [ ] Ensure phrases like `%s ago` and `%s from now` are correctly translated.  
  Compare with https://github.com/WordPress/wordpress-develop/blob/de73a6a8e9a6e2546e0d3624c2cb0daf4d96a819/src/wp-includes/script-loader.php#L140-L167 and https://github.com/WordPress/wordpress-develop/blob/de73a6a8e9a6e2546e0d3624c2cb0daf4d96a819/src/wp-includes/script-loader.php#L326-L364  
  -> **probably new issue**

## User-facing changes

N/A

## Testing Instructions

- Editing media in the editor should still show the creation date.
- Dates in the dashboard should still be shown correctly.
- `wp-includes/js/dist/date.min.js` is no longer loaded in the editor

---

Fixes #4095

### CODE REVIEW

File: assets/src/date/test/getRelativeDisplayDate.js [hunk 1]
```diff
@@ -24,10 +24,11 @@ import MockDate from 'mockdate';
  * Internal dependencies
  */
 import { getRelativeDisplayDate } from '../';
+import { getMoment } from '../getMoment';
```
Comment (swissspidy) [inline assets/src/date/test/getRelativeDisplayDate.js]:
```suggestion
```

**Comment by merapi** [comment]
@googlebot I consent.

**Comment by BrittanyIRL** [comment]
>  it's cumbersome to pass timezone information every time I want to format a date. There should be a way to set a default using something like setSettings.

I agree it's kinda tedious to pass that info. 
For context, the reason I'd opted for that `dateSettings` prop with timezone  in it to the util is to make testing and storybook more straight forward to maintain instead of having to wrap more presentational components in providers. It also reinforced that it's a date util and doesn't tie the structure of the app and components to the date formatting.

**Comment by BrittanyIRL** [comment]
> Ensure phrases like %s ago and %s from now are correctly translated.

Another option is leaning into moment's robust i18n support - https://momentjs.com/docs/#/i18n/

**Comment by swissspidy** [comment]
> Another option is leaning into moment's robust i18n support - momentjs.com/docs/#/i18n

This would mean shipping all of moment's locale files with the plugin, or at least all the locales that are supported by WordPress. There is probably not an exact match between the two systems, plus it would unnecessarily increase plugin size. We would also not have control over the translations, and would be dependant on moment to update them.

We only need to pass a few bits information, and most of that is already available from the WP system. So this is IMO the better way to proceed.

File: assets/src/date/isToday.js [hunk 2]
```diff
@@ -28,8 +33,8 @@ import moment from 'moment-timezone';
  * @return {boolean} If date matches today it will be true
  */
 export function isToday(date) {
-  const displayDate = moment.isMoment(date) ? date : moment.parseZone(date);
-  const today = moment().startOf('day');
+  const displayDate = getMoment(date);
+  const today = moment.utc().startOf('day');
```
Comment (BrittanyIRL) [inline assets/src/date/isToday.js:37]: 👍  this, i think, fixes something that I totally didn't notice on the first round of date utils for formatting.  Was doing some testing with manipulating modified dates for stories w/ different wordPress timezones + setting my browser timezone with an extension to be different and this branch fixes an issue I just saw when testing `main` for a sanity check where when modified date is today the browser date was getting in the way of the relative time formatted. So this is great!  

Dates are always tricky, dates in browser PLUS dates in wordPress is extra tricky.

**Review by BrittanyIRL — COMMENTED** [review]
Spent some time testing this in the dashboard (where I have context for dates) and this works really beautifully.  

My only concern is the failing unit tests - it seems like the tests aren't quite all getting the same date set/looked at. On first look, I can't tell if it's just a setting in the test pipeline that's missed or if something in this update is in the crosshairs.

**Comment by swissspidy** [comment]
@BrittanyIRL It doesn't seem like there are any failing tests right now? Can you perhaps elaborate a bit?

**Comment by BrittanyIRL** [comment]
> @BrittanyIRL It doesn't seem like there are any failing tests right now? Can you perhaps elaborate a bit?

When i run `npm run test:js` locally, the `getRelativeDisplayDate` tests are failing.

**Comment by merapi** [comment]
> When i run `npm run test:js` locally, the `getRelativeDisplayDate` tests are failing.

Fixed them (not getting into details, but CI timezone was different than your local?).

**Comment by BrittanyIRL** [comment]
> > When i run `npm run test:js` locally, the `getRelativeDisplayDate` tests are failing.
> 
> Fixed them (not getting into details, but CI timezone was different than your local?).

probably, but i couldn't possibly be the only one who'd run into this! this team spans _alot_ of timezones. 

thanks for the fix, works great.

**Comment by swissspidy** [comment]
@BrittanyIRL @merapi If one of you wants to take over and add some tests for the new/changed stuff, feel free! Otherwise I'll get to it after my time off.

The other two to-dos warrant new tickets.

**Comment by swissspidy** [comment]
@merapi @BrittanyIRL reviews welcome!

**Comment by swissspidy** [comment]
Ugh, tests pass locally but not on CI. Not sure where the issue is exactly 🤔

File: assets/src/date/test/formatTime.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  beforeEach(() => {
+    MockDate.set(moment.parseZone('2020-07-15T22:47:26+00:00').format());
+  });
+
+  it('should return 12:47 pm with no formatting options', () => {
+    const dateString = '2020-05-02T10:47:26';
+    const formattedTime = formatTime(dateString);
+
+    expect(formattedTime).toBe('12:47 pm');
+  });
```
Comment (barklund) [inline assets/src/date/test/formatTime.js:38]: This one also fails for me:

```
● formatTime › should return 12:47 pm with no formatting options

    expect(received).toBe(expected) // Object.is equality

    Expected: "12:47 pm"
    Received: "6:47 am"

      35 |     const formattedTime = formatTime(dateString);
      36 |
    > 37 |     expect(formattedTime).toBe('12:47 pm');
         |                           ^
      38 |   });
      39 |
      40 |   it('should return 12:47 PM with g:i A formatting options', () => {

      at Object.<anonymous> (assets/src/date/test/formatTime.js:37:27)
```

So something is clearly wrong with timezones somewhere.

**Comment by swissspidy** [comment]
@BrittanyIRL any idea about the timezone issues here?  Feels like I am currently not seeing the forest for the trees.

Edit: nvm. Looks like  the php date formatter takes a JS  date object and then messes everything up with that.

**Comment by BrittanyIRL** [comment]
> @BrittanyIRL any idea about the timezone issues here? Feels like I am currently not seeing the forest for the trees.
> 
> Edit: nvm. Looks like the php date formatter takes a JS date object and then messes everything up with that.

Sorry this pr got lost in my hustle to get keyboard accessibility through the entire dashboard. I will pull and test again today. Are you wanting this in for the release?

**Comment by swissspidy** [comment]
@BrittanyIRL No worries. I think I've found the culprit, and after chatting with Marcin we realized that we should just ditch php-date-formatter. I'll take a stab at it. Nothing for RC though. But maybe a quick win for stable.

**Comment by swissspidy** [comment]
Closing in favor of #4484

**Pull request closed without merging**